From 6cb3ec428b32c402502b86c4bdf336b669c5f6dd Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 19 Jan 2024 10:16:13 +0800
Subject: [PATCH 03/41] kmssink: Fix cropping for video with non-square aspect
 ratio

Give the original video source size to DRM instead of the scaled one.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 sys/kms/gstkmssink.c | 25 ++++++++++++-------------
 1 file changed, 12 insertions(+), 13 deletions(-)

diff --git a/sys/kms/gstkmssink.c b/sys/kms/gstkmssink.c
index 59ed881..6330e88 100644
--- a/sys/kms/gstkmssink.c
+++ b/sys/kms/gstkmssink.c
@@ -2137,15 +2137,18 @@ gst_kms_sink_show_frame (GstVideoSink * vsink, GstBuffer * buf)
   if (buf) {
     buffer = gst_kms_sink_get_input_buffer (self, buf);
     vinfo = &self->vinfo;
-    video_width = src.w = GST_VIDEO_SINK_WIDTH (self);
-    video_height = src.h = GST_VIDEO_SINK_HEIGHT (self);
+    src.w = GST_VIDEO_SINK_WIDTH (self);
+    src.h = GST_VIDEO_SINK_HEIGHT (self);
   } else if (self->last_buffer) {
     buffer = gst_buffer_ref (self->last_buffer);
     vinfo = &self->last_vinfo;
-    video_width = src.w = self->last_width;
-    video_height = src.h = self->last_height;
+    src.w = self->last_width;
+    src.h = self->last_height;
   }
 
+  video_width = GST_VIDEO_INFO_WIDTH (vinfo);
+  video_height = GST_VIDEO_INFO_HEIGHT (vinfo);
+
   /* Make sure buf is not used accidentally */
   buf = NULL;
 
@@ -2166,8 +2169,8 @@ gst_kms_sink_show_frame (GstVideoSink * vsink, GstBuffer * buf)
   if ((crop = gst_buffer_get_video_crop_meta (buffer))) {
     GstVideoInfo cropped_vinfo = *vinfo;
 
-    cropped_vinfo.width = crop->width;
-    cropped_vinfo.height = crop->height;
+    video_width = src.w = cropped_vinfo.width = crop->width;
+    video_height = src.h = cropped_vinfo.height = crop->height;
 
     if (!gst_kms_sink_calculate_display_ratio (self, &cropped_vinfo, &src.w,
             &src.h))
@@ -2186,13 +2189,9 @@ retry_set_plane:
   result.x += self->render_rect.x;
   result.y += self->render_rect.y;
 
-  if (crop) {
-    src.w = crop->width;
-    src.h = crop->height;
-  } else {
-    src.w = video_width;
-    src.h = video_height;
-  }
+  /* Restore the real source size */
+  src.w = video_width;
+  src.h = video_height;
 
   /* handle out of screen case */
   if ((result.x + result.w) > self->hdisplay)
-- 
2.39.5

